kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ChecksumReader.java

```diff
@@ -0,0 +1,129 @@
+import static io.delta.kernel.internal.util.FileNames.isChecksumFile;
+import static io.delta.kernel.internal.util.Utils.singletonCloseableIterator;
+
+import io.delta.kernel.data.ColumnarBatch;
+import io.delta.kernel.engine.Engine;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.util.FileNames;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.FileStatus;
+import java.util.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Utility method to load protocol and metadata from the Delta log checksum files. */
+public class ChecksumReader {
+  private static final Logger logger = LoggerFactory.getLogger(ChecksumReader.class);
+
+  /**
+   * Load the protocol and metadata from the checksum file at the given version. If the checksum
+   * file is not found at the given version, it will try to find the latest checksum file that is
+   * created after the lower bound version or within the last 100 versions.
+   *
+   * @param engine the engine to use for reading the checksum file
+   * @param logPath the path to the Delta log
+   * @param readVersion the version to read the checksum file from
+   * @param lowerBoundOpt the inclusive lower bound version to search for the checksum file
+   * @return Optional {@link CRCInfo} containing the protocol and metadata, and the version of the
+   *     checksum file. If the checksum file is not found, it will return an empty
+   */
+  public static Optional<CRCInfo> getCRCInfo(
+      Engine engine, Path logPath, long readVersion, Optional<Long> lowerBoundOpt) {
+    logger.info("Loading CRC file for version {}", readVersion);
+    // First try to load the CRC at given version. If not found or failed to read then try to
+    // find the latest CRC file that is created after the lower bound version or within the last 100
+    // versions if no lower bound is provided.
+    Path crcFilePath = checksumFile(logPath, readVersion);
+    Optional<CRCInfo> crcInfoOpt = readChecksumFile(engine, crcFilePath);
+    if (crcInfoOpt.isPresent()
+        ||
+        // we don't expect any more checksum files as it is the first version
+        readVersion == 0) {
+      return crcInfoOpt;
+    }
+
+    // Try to list the last 100 CRC files and see if we can find a CRC that we can use
+    long lowerBound = Math.max(lowerBoundOpt.orElse(0L), Math.max(0, readVersion - 100));
+    logger.info(
+        "CRC file for version {} not found, attempt to loading version up to {}",
+        readVersion,
+        lowerBound);
+
+    Path lowerBoundFilePath = checksumFile(logPath, lowerBound);
+    try (CloseableIterator<FileStatus> crcFiles =
+        engine.getFileSystemClient().listFrom(lowerBoundFilePath.toString())) {
+      List<FileStatus> crcFilesList = new ArrayList<>();
+      crcFiles
+          .filter(file -> isChecksumFile(new Path(file.getPath())))
+          .forEachRemaining(crcFilesList::add);
```

Comment:
@scottsand-db I wonder if this can be refactored to use any of the log listing methods in your PR?